src/objects.c

```diff
@@ -1084,13 +1089,86 @@ static P11PROV_OBJ *find_associated_obj(P11PROV_CTX *provctx, P11PROV_OBJ *obj,
 
 static void p11prov_obj_refresh(P11PROV_OBJ *obj)
 {
+    int login_behavior;
+    bool login = false;
+    CK_SLOT_ID slotid = CK_UNAVAILABLE_INFORMATION;
+    P11PROV_SESSION *session = NULL;
+    CK_SESSION_HANDLE sess = CK_INVALID_HANDLE;
+    CK_ATTRIBUTE template[3] = { 0 };
+    CK_ATTRIBUTE *attr;
+    int anum;
+    CK_OBJECT_HANDLE handle;
+    CK_ULONG objcount = 0;
     P11PROV_OBJ *tmp = NULL;
-    tmp = find_associated_obj(obj->ctx, obj, obj->class);
-    if (!tmp) {
-        /* nothing we can do, invalid handle it is */
+    CK_RV ret;
+
+    P11PROV_debug("Refresh object %p", obj);
+
+    if (obj->class == CKO_PRIVATE_KEY) {
+        login = true;
+    }
+    login_behavior = p11prov_ctx_login_behavior(obj->ctx);
+    if (login_behavior == PUBKEY_LOGIN_ALWAYS) {
+        login = true;
+    }
+
+    slotid = p11prov_obj_get_slotid(obj);
+
+    ret = p11prov_get_session(obj->ctx, &slotid, NULL, obj->refresh_uri,
+                              CK_UNAVAILABLE_INFORMATION, NULL, NULL, login,
+                              false, &session);
+
+    if (ret != CKR_OK) {
+        P11PROV_debug("Failed to get session to refresh object %p", obj);
         return;
     }
 
+    sess = p11prov_session_handle(session);
+
+    anum = 0;
+    CKATTR_ASSIGN(template[anum], CKA_CLASS, &(obj->class), sizeof(obj->class));
+    anum++;
+    /* use CKA_ID if available */
+    attr = p11prov_obj_get_attr(obj, CKA_ID);
+    if (attr) {
+        template[anum] = *attr;
+        anum++;
+    }
+    /* use Label if available */
+    attr = p11prov_obj_get_attr(obj, CKA_LABEL);
+    if (attr) {
+        template[anum] = *attr;
+        anum++;
+    }
+
+    ret = p11prov_FindObjectsInit(obj->ctx, sess, template, anum);
+    if (ret != CKR_OK) {
+        goto done;
+    }
+
+    /* we expect a single entry */
+    ret = p11prov_FindObjects(obj->ctx, sess, &handle, 1, &objcount);
+
+    /* Finalizing is not fatal so ignore result */
+    p11prov_FindObjectsFinal(obj->ctx, sess);
+
+    if (ret != CKR_OK) {
+        P11PROV_debug("Failed to find refresh object %p", obj);
+        goto done;
+    }
+    if (objcount != 1) {
```

Comment:
Maybe we should distinguish the cases  objcount = 0 and objcount > 1 here.
And if objcount > 1, should we pick the first one? (Although we expect a single entry?)

Reply:
No, if we pick the wrong one we could cause cryptographic operations to happen using the wrong key, that would be CVE worthy material.